base_layer/mmr/src/sparse_merkle_tree/node.rs

```diff
@@ -110,7 +109,72 @@ hash_type!(NodeHash);
 hash_type!(ValueHash);
 hash_type!(NodeKey);
 
-pub const EMPTY_NODE_HASH: NodeHash = NodeHash([0; 32]);
+impl NodeKey {
+    pub fn as_directions(&self) -> PathIterator {
+        PathIterator::new(self)
+    }
+}
+
+pub const EMPTY_NODE_HASH: NodeHash = NodeHash([0; KEY_LENGTH]);
+
+pub struct PathIterator<'a> {
+    cursor_front: usize,
+    // position *after* next bit when going backwards
+    cursor_back: usize,
+    key: &'a NodeKey,
+}
+
+impl PathIterator<'_> {
+    pub fn new(key: &NodeKey) -> PathIterator {
+        PathIterator {
+            cursor_front: 0,
+            // KEY_LENGTH is currently 32 bytes, so this will not overflow
+            cursor_back: KEY_LENGTH * 8,
+            key,
+        }
+    }
+}
+
+impl<'a> Iterator for PathIterator<'a> {
+    type Item = TraverseDirection;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.cursor_front >= self.cursor_back {
+            return None;
+        }
+        let bit = get_bit(self.key.as_slice(), self.cursor_front);
+        self.cursor_front += 1;
+        Some(bit_to_dir(bit))
+    }
+
+    // This must be overridden, otherwise iterator connectors don't work
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.cursor_back.saturating_sub(self.cursor_front);
+        (len, Some(len))
+    }
+}
+
+impl<'a> DoubleEndedIterator for PathIterator<'a> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        if self.cursor_front >= self.cursor_back {
+            return None;
+        }
+        self.cursor_back -= 1;
+        let bit = get_bit(self.key.as_slice(), self.cursor_back);
+        Some(bit_to_dir(bit))
+    }
+
+    fn nth_back(&mut self, n: usize) -> Option<Self::Item> {
+        self.cursor_back -= n;
```

Comment:
Could underflow. Should check for this case and return None.

Reply:
Good catch